presto-hive/src/test/java/com/facebook/presto/hive/TestS3SelectRecordCursor.java

```diff
@@ -0,0 +1,278 @@
+    {
+        String ddlSerializationValue = "struct article { int article, double author, xxxx date_pub, int quantity}";
+        String expectedDDLSerialization = "struct article { int article, double author, xxxx date_pub, int quantity}";
+        assertEquals(buildSplitSchema(ddlSerializationValue, DEFAULT_TEST_COLUMNS),
+                buildExpectedProperties(expectedDDLSerialization, DEFAULT_TEST_COLUMNS));
+    }
+
+    @Test
+    public void shouldOnlyGetColumnTypeFromHiveObjectAndNotFromDDLSerialLastColumnNameWithoutEndStruct()
+    {
+        String ddlSerializationValue = "struct article { int article, double author, xxxx date_pub, int quantity }";
+        String expectedDDLSerialization = "struct article { int article, double author, xxxx date_pub, int quantity}";
+        assertEquals(buildSplitSchema(ddlSerializationValue, DEFAULT_TEST_COLUMNS),
+                buildExpectedProperties(expectedDDLSerialization, DEFAULT_TEST_COLUMNS));
+    }
+
+    @Test(expectedExceptions = NullPointerException.class)
+    public void shouldThrowNullPointerExceptionWhenColumnsIsNull()
+    {
+        updateSplitSchema(new Properties(), null);
+    }
+
+    @Test(expectedExceptions = NullPointerException.class)
+    public void shouldThrowNullPointerExceptionWhenSchemaIsNull()
+    {
+        updateSplitSchema(null, ImmutableList.of());
+    }
+
+    private Properties buildSplitSchema(String ddlSerializationValue, HiveColumnHandle... columns)
+    {
+        Properties properties = new Properties();
+        properties.put(SERIALIZATION_LIB, LAZY_SERDE_CLASS_NAME);
+        properties.put(SERIALIZATION_DDL, ddlSerializationValue);
+        return updateSplitSchema(properties, asList(columns));
+    }
+
+    private Properties buildExpectedProperties(String expectedDDLSerialization, HiveColumnHandle... expectedColumns)
+    {
+        String expectedColumnsType = getTypes(expectedColumns);
+        String expectedColumnsName = getName(expectedColumns);
+        Properties propExpected = new Properties();
+        propExpected.put(LIST_COLUMNS, expectedColumnsName);
+        propExpected.put(SERIALIZATION_LIB, LAZY_SERDE_CLASS_NAME);
+        propExpected.put(SERIALIZATION_DDL, expectedDDLSerialization);
+        propExpected.put(LIST_COLUMN_TYPES, expectedColumnsType);
+        return propExpected;
+    }
+
+    private String getName(HiveColumnHandle[] expectedColumns)
+    {
+        return Stream.of(expectedColumns)
+                .map(HiveColumnHandle::getName)
+                .collect(joining(","));
+    }
+
+    private String getTypes(HiveColumnHandle[] expectedColumns)
+    {
+        return Stream.of(expectedColumns)
+                .map(HiveColumnHandle::getHiveType)
+                .map(HiveType::getTypeInfo)
+                .map(TypeInfo::getTypeName)
+                .collect(joining(","));
+    }
+
+    private static final RecordReader MOCK_RECORD_READER = new RecordReader() {
```

Comment:
Nit: opening brace should be on the next line, since this is defining a class

Reply:
`RecordReader` is a parameterized class, so this should have generics:
```java
private static final RecordReader<Object, Object> MOCK_RECORD_READER = new RecordReader<>()
```
Or perhaps
```java
private static final RecordReader<?, ?> MOCK_RECORD_READER = new RecordReader<>()
```
I would need to open this in my IDE and see how it's used. But it should not be used as a "raw class".

Reply:
Fixed

```
private static final RecordReader<?, ?> MOCK_RECORD_READER = new RecordReader<>()
```